### PR TITLE
use struct tag "mutable" for validating if a field can change

### PR DESCRIPTION
**Pull request by asalkeld**
not perfect, but a start

### CODE REVIEW

File: pkg/api/v20191231preview/openshiftcluster_validate_test.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+		}
+		for _, k := range in.MapKeys() {
+			mutateAndTest(t, inOc, origOc, in.MapIndex(k), orig.MapIndex(k))
+		}
+	case reflect.Ptr:
+		if in.IsNil() || orig.IsNil() {
+			return
+		}
+		mutateAndTest(t, inOc, origOc, in.Elem(), orig.Elem())
+	case reflect.Slice:
```
Comment (asalkeld) [inline pkg/api/v20191231preview/openshiftcluster_validate_test.go]: so this is why the manual test is still in place above, any bright ideas?
I suspect i'll have to totally restructure the flow to account for this

**Comment by jim-minter** [comment]
@asalkeld have reworked this substantially, ptal?

File: pkg/api/v20191231preview/openshiftcluster_validate_test.go [hunk 2]
```diff
@@ -400,3 +409,116 @@ func TestValidateWorkerProfile(t *testing.T) {
...
+func zeroVal(t reflect.Type) reflect.Value {
+	return reflect.New(t).Elem()
+}
+
+func TestValidateOpenShiftClusterDelta(t *testing.T) {
+	oc, mut := validOpenShiftCluster(), validOpenShiftCluster()
+
+	v := reflect.ValueOf(mut).Elem()
+
+	walk(func(path string, mutable bool) {
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster_validate_test.go]: It seems to me that here we are basicaly testing the `github.com/jim-minter/rp/pkg/util/immutable` package. Can we test the package itself with simpler test cases with known value? Like:

* Check `mutable:"case"` validation
* Check `mutable:"true"` validation
* Check validation of fields without the tag
* Check recursive validation
* etc

Then there will be no need to *unit test* `ValidateOpenShiftClusterDelta` as it only contains `immutable.Validate` call.

IMHO, `walk` looks too complex for unit tests and requires tests itself.

File: pkg/util/immutable/immutable.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+	default:
+		return fmt.Errorf("unimplemented kind %s", v.Kind())
+	}
+
+	return nil
+}
+
+func validationError(path string) error {
+	return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodePropertyChangeNotAllowed, path, fmt.Sprintf("Changing property '%s' is not allowed.", path))
```
Comment (m1kola) [inline pkg/util/immutable/immutable.go]: `api.CloudError` doesn't seem to belong here. As an option we can move the package to `pkg/api` that it only must be used with API stuff.

Or we can define own error type here and convert to `api.CloudError` in API?

**Review by m1kola — CHANGES_REQUESTED** [review]
I like how the field tag removes a lot of boilerplate code, but tests seem to be quite complex. I'm wondering, if we could make them simpler?

File: pkg/api/v20191231preview/openshiftcluster_validate_test.go (hunk 2 above)
Comment (jim-minter) [reply pkg/api/v20191231preview/openshiftcluster_validate_test.go]: I'm glad you've got the same concerns I have.

The thing that concerns me about testing solely immutable.Validate is the risk of the actual validation somehow being wrong and not having test coverage there.

I wonder whether the test regime is trying to be too smart and we should just write and maintain all the tests manually and clearly.

But then, I wonder, if we do that for the tests - is worth bothering with immutable.Validate, and maybe we keep the actual functionality dumb as well?

Perhaps we should start with the middle ground - use immutable.Validate but test manually - and reassess later?

Comment (m1kola) [reply pkg/api/v20191231preview/openshiftcluster_validate_test.go]: > IMHO, `walk` looks too complex for unit tests and requires tests itself.

Just to clarify. I don't want us to cover `walk` with tests :) I want to get rid of it by using simplier (mock) structures.

Comment (m1kola) [reply pkg/api/v20191231preview/openshiftcluster_validate_test.go]: > Perhaps we should start with the middle ground - use immutable.Validate but test manually - and reassess later?

Yeah, sounds like a good plan.

File: pkg/util/immutable/immutable.go (hunk 3 above)
Comment (asalkeld) [reply pkg/util/immutable/immutable.go]: agree, just have a different error type (maybe 2 one for case and one for immutable) with the field name

Comment (jim-minter) [reply pkg/util/immutable/immutable.go]: I agree with the principle, but I see no need to actually change anything.

**Comment by jim-minter** [comment]
merged, thanks

**Pull request closed without merging**